### DISCUSSION TITLE
Changes in Records Data Storage Format

### DISCUSSION THREAD

**Discussion by 0xc0de4c0ffee** [namesys-eth/ccip2-eth-resources#6] [Ideas]
Current design is to handle records.json as `domain.eth.limo/.well-known/eth/domain/records.json` and internally route contenthash so domain will resolve for other records. 

I was testing possible new format to handle records.json, with some big upsides and small down side. 
Update : 
1) use `<ipns-hash>.ipns.dweb.link/eth/domain/records.json` and `dweb.link/ipns/<hash>/eth/domain/<records>.json` format
2) contenthash is handled at `../eth/domain/contenthash.json` for `domain.eth.limo`
3) if there's non-ens `.well-known` records, it'll will be in `domain.eth.limo/.well-known/..`

+ve side: easy to manage records.
1) we can add and randomize public gateways instead of sending all req to eth.limo and expect fix to fallback bug soon.
2) we get clean IPNS and don't have to handle "IPFS" directory structures in IPNS publish as it's standalone now.
3) we can run IPFS + IPNS services to manage all records easily as we don't have to directly deal with html/js inside IPFS directory.
4) we don't break anything in IPNS+IPLD or ?require future changes to handle IPLD+IPNS+ENS.

-ve side: same as current design problem.
1) web pages and other content and LN/Nostr related .well-known records have to go inside contenthash. we don't manage that directly, only keep that hash in json.
2) .. * have to finish build & test to find more

> domain.eth >> IPNS >> CCIP read >> resolve record

Still WIP idea so we've to build both to compare full differences.

**Comment by sshmatrix** [comment]
> 1. use `<ipns-hash>.ipns.dweb.link/eth/domain/records.json` and `dweb.link/ipns/<hash>/eth/domain/<records>.json` format

Ok, so not `.well-known` directory for ENS stuff.

> 2. contenthash is handled at `../eth/domain/contenthash.json` for `domain.eth.limo`

Ok, but I thought contenthash was on-chain? Or, you are proposing that the contenthash is also kept off-chain and then queried by CCIP-Read?

> 2. we get clean IPNS and don't have to handle "IPFS" directory structures in IPNS publish as it's standalone now.
> 3. we can run IPFS + IPNS services to manage all records easily as we don't have to directly deal with html/js inside IPFS directory.

I didn't understand how we are bypassing "IPFS". Is it possible to add content to an IPNS key directly?

**Reply by 0xc0de4c0ffee** [reply]
bit recap to explain issues/ 

1) owner of domain.eth sets resolver to ccip2.eth

2) owner generates IPNS keys & set 1 IPNS hash "0xe5010172...", libp2p key
a)  encode base36 (3.a) for every ccip-read
b) OR use simpler hex format string("f+0172...") * by removing `e501`, better than (2.a)  but we lose (3.a) \<sub\>.ipns.. feature. 

3) Resolver is using public IPFS gateways to read all ENS records.json.
a) `<k+cid-v1-base36>.ipns.gateway.tld/...records.json` OR
b) OR `gateway.tld/ipns/f<contenthash[2:]>/...records.json` \* f+hexString

4) owner can sign same/another IPNS/IPFS hash for contenthash.json to resolve web contents only.
a) domain.eth.limo >resolves> contenthash.json, ?2nd fallbacks to on-chain (+fallback bug)
b) all other ENS records can be still resolved using ENS as it's reading from public gateways
c) domain.eth.limo/.well-known/.. files for non ens records will be resolved from contenthash.json 1st. 

why? 
it's simplifying whole process & separating ENS records.json from web contents.
we don't have to deal with html/big contents directly to update ENS records, jus add a simple index.html for default profile. 
& .well-known/.. works out of box too, we've to set same ipns hash as default signed data in contenthash.json,
Users have to copy-paste that dir only (not ENS records) if they change their contenthash.json to publish new web content using 2nd IPNS/IPFS keys.

**Reply by sshmatrix** [reply]
Okay, I think I got the gist of what you are saying, although I am still slightly lost on the specific sub-points related to implementation. I will try to re-read/process it. In the end, we don't want people to worry about remembering to keep their records in their web content whenever they want to update their contenthash. Basically separate the records from the web content. Correct me if I am wrong

**Reply by sshmatrix** [reply]
Ok, I got what you mean. We should implement this. It makes total sense to separate the web content (`eth.limo/`) and the records (`.ipns.dweb.link`) with two independent gateways. Only one issue: this means that we will need to keep a separate IPNS key record in CCIP2 Resolver for the records.json. In other words, when changing the Resolver to ours, users will also be keygen + deploying a new IPNS record inside the new Resolver, correct?

**Reply by sshmatrix** [reply]
> 3. if there's non-ens `.well-known` records, it'll will be in `domain.eth.limo/.well-known/..`

Why not **also** `<ipns-hash>.ipns.dweb.link/.well-known/..`? I know it is less useful since users will surely prefer the cleaner gateway for real-world use but doesn't hurt to provide the feature? Or feature-creeping? 😝

**Reply by 0xc0de4c0ffee** [reply]
Users have one IPNS contenthash set on-chain resolver: 
a) default : add same ipns in .../contenthash.json, it'll be used for webcontent + /.well-known/..
b) if they use different IPNS/IPFS contenthash in json, on-chain CH is used for ENS records only & new CH is used for dweb including /.well-known/..

Most users won't need 2nd CH, it's for advance users/dapps, & all future safe for IPLD/.../records.json as we don't want to add html/js/.. in ipld then work on more specs to resolve those plaintext/datauris.

**Reply by sshmatrix** [reply]
Ok, I see. That sounds good 👌🏻

**Comment by 0xc0de4c0ffee** [comment]
I've updated https://github.com/namesys-eth/ccip2-eth-resolver/pull/3 
It's still missing few functions to manage stuffs around,  >80% ready + missing full test.

@sshmatrix / @tomlightning / @eth-limo  🙏  if you've suggestions/ questions tag me on PR or here.

**Reply by sshmatrix** [reply]
Looks like the client (protoype: https://namesys-eth.github.io/ccip2-eth-client/) and contracts will be ready simultaneously 🍾
